### DISCUSSION TITLE
Changes to Push

### DISCUSSION THREAD

**Discussion by aritchie** [shinyorg/shiny#726] [Ideas]
I am considering chopping everything to do with the notification in Shiny Push.  I have had to repeatedly explain to people that this data is not present in the background in any case.  I also frequently have to explain the use of content-available and the necessity of including actual data parameters being sent in order to engage the background event. 

Add the complexity of channels /w responses and additional cases like multilingual which just causes things to get out of control.  Anytime I have ever had this type of complexity, I switch off to local notifications in any case.  As such, I am considering going to a simplified model:

```csharp
public interface IPushDelegate 
{
    Task OnReceived(IReadOnlyDictionary<string, string> data);
}
```

If you need to do anything with notifications, you can either allow the OS to do most of it, but if you have the complex case of notification responses, let Shiny Notifications do this for you.

```csharp
// your startup
services.UseNotification<MyNotificationDelegate>();
services.UsePushWhatever<MyPushDelegate>();
services.AddSingleton<ILocalize, YourLocalizeComponent>();


// your push delegate
public class MyPushDelegate : IPushDelegate 
{
   readonly INotificationManager notifications;
   readonly ILocalize localize; 

   public MyPushDelegate(INotificationManager notifications, ILocalize localize) {
      this.notifications = notifications;
      this.localize = localize;
   }

   public async Task OnReceived(IReadOnlyDictionary<string, string> data) 
   {
      if (data.ContainKeys("Event")) 
     {
        await this.notifications.Send(this.localize["MyKey"]); // add things like channel here if necessary
     }
   }
}


// your notification delegate
public class MyNotificationDelegate : INotificationDelegate
{
   public async Task OnEntry(NotificationResponse response) 
   {
      // do what you want with the shiny notification here
   }
}
```

NOTE that Shiny Notifications specifically ignores push notifications by design and will continue to do so, thus you would have to send the notification like shown above to get the delegate to fire.

**Comment by Codelisk** [comment]
Sounds good, the only issue I see with this is when for example I want to track when a critical alert in ios was pressed. For my understanding i can not send a critical alert as a silent notification and have to use the "normal" notification, and then it would not be possible to track when its pressed.

**Reply by aritchie** [reply]
You can send critical notifications with the local api on iOS as well.  You can track entry of the notification with that.

**Reply by Codelisk** [reply]
Oh okay thank you I will look into that!

**Comment by GuidoNeele** [comment]
I've looked at this myself (before posting issue #656) but decided against it because it looked like I couldn't rely on APNs to always deliver the push notification (https://stackoverflow.com/questions/53531369/limits-on-silent-push-notifications-per-hour). Is this a solution that can be used in all scenarios?

**Reply by aritchie** [reply]
@GuidoNeele This is a good point - thank you.  Good enough that I'll cancel out this set of changes
